Add first files for the Kubernetes Short Course

jonatasbaldin opened the pull request:
## Description:
Hi there! This is my first PR on the syllabus 🥳  

I'm adding a first draft for the Kubernetes course, just a bit of content to see if I got the process right.

I have a couple of questions:
- Is the folder structure ok?
- I used number conventions (1.0, 1.1) on the `title` field to organize the content in order. Should it be done like this? Is there a better way?
- Is there a guideline to assign/label a PR like this?

## I solemnly swear that:

- [x] I ran the hugo server and looked at my changed in the browser with my own eyes
- [x] I ran the linter and there were no errors
- [x] My code follows the style guidelines of this project **(I guess!)**
- [x] I have performed a self-review of my own code

sheenarbw commented on content/k8s/bare-bones-deployment/project-overview/_index.md:
```diff
@@ -0,0 +1,16 @@
+---
+title: 1.0 Project Overview
+content_type: topic
```
Where is the actual project?

Learners can move topic cards into the complete columns themselves. Project cards represent deliverables.

It looks like there are no deliverables at all

jonatasbaldin replied:
For the first module, I added on project, `content/k8s/manual-app-deployment/nginx-reverse-proxy`.

sheenarbw commented on content/k8s/bare-bones-deployment/setup-ec2/_index.md:
```diff
@@ -0,0 +1,28 @@
+
+TODO: Content to be reviewed if we automate the EC2 instance creation.
+
+**Create a Key Pair:**
```
Please use markfdown formatting for text.

Headings use # symbols. Then that gets converted into the correct html tags.

jonatasbaldin replied:
We just removed this content for now.

sheenarbw commented on content/k8s/bare-bones-deployment/setup-ec2/_index.md:
```diff
@@ -0,0 +1,28 @@
+ready: false
+---
+
+TODO: Content to be reviewed if we automate the EC2 instance creation.
```
Is this meant to be done? I'm not sure it's ready

jonatasbaldin replied:
We just removed this content for now.

sheenarbw commented on content/k8s/introduction/_index.md:
```diff
@@ -0,0 +1,46 @@
+---
+
+## What is Kubernetes?
+In today's world of complex, web-scale application backends made up of many microservices and components running across clusters of servers and containers, managing and coordinating all these pieces is incredibly challenging.
```
Do the learners know what microservices and containers are at this point? 

I wonder if there are any useful explanations you can link to. I'm sure there are videos and docs on the internet that expect less foundational knowledge.

sheenarbw commented on content/k8s/introduction/_index.md:
```diff
@@ -0,0 +1,46 @@
+---
+title: 0. Introduction
```
Nothing has this introduction as a prerequisite. That means that learners can skip this card and do it at the end.

jonatasbaldin replied:
Added it as prerequisite to the first chapter on the second module.

sheenarbw commented on content/k8s/bare-bones-deployment/project-overview/_index.md:
```diff
@@ -0,0 +1,16 @@
+3. Develop a Flask API that handles client requests and provides services over HTTPS. [https://flask.palletsprojects.com/en/3.0.x/](https://flask.palletsprojects.com/en/3.0.x/)
+4. Set up a reverse proxy within Nginx pointing to your Flask app, ensuring smooth communication.
+
+By integrating these components, your project will offer secure client access, host a Python (Flask) API, and maintain stable performance in various environments.
```
What is the learner meant to do? How do they submit their project? 

Always imagine what the learner will ask and try to answer that up front otherwise you'll end up answering the same quesion over and over manually.

I assume you want the learners to submit a repo with their yaml files. Do you also want them to submit a link? 

Be explicit.

This will also make the syllabus reusable. If other staff members are not sure what is expected then it means people will get inventive.

jonatasbaldin replied:
I added two lines regarding the deliverables, but I want to circle back with @t-witchy and @RLutsch if that's accurate.

sheenarbw commented on content/k8s/bare-bones-deployment/certbot/_index.md:
```diff
@@ -0,0 +1,39 @@
+ready: true
+---
+
+To be able to get a TLS certificate for HTTPS we will install certbot and run in certonly mode where it will generate the certificates. Take note of the paths it spits out where it stores the certificate, it should something like:
```
Do the learners know what TLS is at this point? Is there a resource we can link to that assumes less prior knowledge?

jonatasbaldin replied:
I added a bit of introduction to TLS and an external link for more information.

sheenarbw reviewed (changes requested):
This is a good start, you're definately on the right track.

There are 2 things missing:

1. There should be a main syllabus outline file that lists the content in the right order. This will be used to figure out what cards should show up on people's boards. Look at https://github.com/Umuzi-org/ACN-syllabus/tree/develop/content/syllabuses to see examples. You can put your syllabus outline in there
2. there are no projects specified, only topics

I'm also not super sure that everyone will be able to keep up with this content. It assumes  a lot of prior knowledge. It's nice to find explanitry content that other people maintain and link to that. You're used to this stuff, the learners are noobs and they are likely to feel overwhelmed if you throw too many new words at them without explaining nicely.

It's also useful to be extra explicit about what is required from the learners at different points.

To answer your questions:
- your folder structure is fine
- we dont have a tagging system for prs. We usually keep track of stuff in a seperate project management tool so that means we can get away with being a little messy
- the titles look ok

The final order of the cards gets calculated based on what is in the main syllabus file and the prerequisites of each piece of content. 
The learners are allowed to do things out of order if things dont have prerequisites set up.  

The system does not order things by title or anything like that.

jonatasbaldin commented:
Hi @sheenarbw, thanks for the review and the comments.

This is the first time we are touching on the Syllabus and Tilde, so this PR might get a bit messy until we get comfortable with both projects, hope you understand it 😅

I removed the number from the titles. The order doesn't get correct on the main Syllabus page, unfortunately, but it will be correct on Tilde based on the prerequisites metadata. If we decide the pages should be ordered in Syllabus, we can add the ordering numbers back.

I added our first module, `Manual App Deployment`, and it has only one `project` with a deliverable, at the end. As I understood, only the `submission_type` is required, right?

We will work to bring more context into the content ASAP.

sheenarbw commented:
Hey @jonatasbaldin 

If I'm ever slow to respond here please forgive me, I get a lot of spam from learners so my github notifications are a complete dog show. If you need to ask questions then the best place to get my attention is slack.

You can just add the submission type for projects unless it's a `continue_repo` project. These are used for multi-part projects where you want a learner to add to a single repo over the course of many projects. This is useful for big projects because then we can track progress over time and give people guidance and feedback on different parts.

The challenge with `continue_repo` projects is that if a person hasn't completed part 1 (as in, they have not been marked as competent) then they are blocked from starting part 2. In these cases, it's useful to have other content for the learner to work on while they wait for feedback.

jonatasbaldin commented on content/k8s/manual-app-deployment/ec2-setup/index.md:
```diff
@@ -0,0 +1,30 @@
+---
```
Since we will automate the instance creation process, we can modify this page.

Instead of "Setup EC2", we can a have page named "Accessing your EC2 instance", where we tell them how to:
- Save the `key.pem` file with the content given by us
- Change the `key.pem` file permissions (if needed)
- Use SSH to access their instance

We also need to make this OS agnostic, maybe they have Windows/Linux/OSX.

What do you think @t-witchy?

jonatasbaldin commented on content/k8s/manual-app-deployment/postgresql-installation/_index.md:
```diff
@@ -5,7 +5,8 @@ tags:
 prerequisites:
-  hard: []
+  hard: 
+  - k8s/manual-app-deployment/nginx-installation
```
They actually don't need to have Nginx before PostgreSQL, they can do this in a different order.

Maybe flag as as soft prereq? Or remove the prereqs?

jonatasbaldin commented on content/k8s/manual-app-deployment/nginx-installation/_index.md:
```diff
@@ -5,7 +5,8 @@ tags:
 prerequisites:
-  hard: []
+  hard: 
+  - k8s/manual-app-deployment/certbot
```
They actually don't need to have Certbot before installing Nginx, they can do this in a different order.

Maybe flag as as soft prereq? Or remove the prereqs?

jonatasbaldin commented on content/k8s/manual-app-deployment/github-setup/_index.md:
```diff
@@ -5,7 +5,8 @@ tags:
 prerequisites:
-  hard: []
+  hard:
+  - k8s/manual-app-deployment/setup-ec2
```
They actually don't need to have GitHub before having the EC2, they can do this in a different order.

Maybe flag as as soft prereq? Or remove the prereqs?

jonatasbaldin commented:
Hey @sheenarbw! Don't worry about the response time, If I need something more urgent I'll ping you on Slack.

I don't think we will be using `continue_repo`, only `submission_type: link`, which is easier as well.

jonatasbaldin commented on content/syllabuses/k8s.md:
```diff
@@ -0,0 +1,24 @@
+- {{< contentlink path="k8s/containers-with-docker/nginx" >}}
+- {{< contentlink path="k8s/containers-with-docker/python-app" >}}
+- {{< contentlink path="k8s/containers-with-docker/running" >}}
+- {{< contentlink path="k8s/kubernetes/deploy-app-with-manifest-file" >}}
```
The correct ordering here is:

```
- {{< contentlink path="k8s/kubernetes/k3s-quickstart" >}}
- {{< contentlink path="k8s/kubernetes/expanding-the-cluster" >}}
- {{< contentlink path="k8s/kubernetes/deploy-app-with-manifest-file" >}}
```

Also, how will the ordering between pages under each module work? Through prerequisites? Or should we make them explicit here?

jonatasbaldin commented on content/k8s/kubernetes/k3s-quickstart/_index.md:
```diff
@@ -0,0 +1,101 @@
+---
```
@t-witchy I think you added this file?

It is missing some content. Before the Deployment, we install K3d and talk about other resources as well.

https://app.gitbook.com/o/2DzlYnPstQTFtiSgav55/s/j3rSnsmwqXaNyuH81o0A/practical-component-chapters/k8s-the-containers/activity-1/3.1.1-k3s-introduction

jonatasbaldin commented on content/k8s/containers-with-docker/certbot/_index.md:
```diff
@@ -6,6 +6,7 @@ from_repo: k8s/manual-app-deployment/project-overview
 prerequisites:
   hard:
   - k8s/containers-with-docker/containers-overview
+  - k8s/manual-app-deployment/project-overview
```
hey there! 

Why is this a prereq for both the whole chapter? Is this for better Tilde integration?

ry-oc replied:
That is the source of the continue repo, so it needs to be a pre-req of every project that uses that repo
